### PR TITLE
Fix sources for version configurations

### LINKED ISSUE #2: grokproject referencing svn.zope.org incorrectly
Hi, I was going to check out Grok, and I am unable to even follow the getting started tutorial as grokproject appears to be broken...  Here is my output:

```(grok-env) kveroneau@debian:~$ grokproject Sample
Enter user (Name of an initial administrator user): grok
Enter passwd (Password for the initial administrator user): 
Determining current grok version...
Creating directory ./Sample
Running /zfs/kveroneau/grok-env/bin/python /zfs/kveroneau/Sample/bootstrap.py...
/zfs/kveroneau/grok-env/lib/python2.7/site-packages/pkg_resources/__init__.py:184: RuntimeWarning: You have iterated over the result of pkg_resources.parse_version. This is a legacy behavior which is inconsistent with the new version class introduced in setuptools 8.0. In most cases, conversion to a tuple is unnecessary. For comparison of versions, sort the Version instances directly. If you have another use case requiring the tuple, please file a bug with the setuptools project describing that need.
  stacklevel=1,
While:
  Initializing.
Error: Error downloading extends for URL http://svn.zope.org/*checkout*/zopetoolkit/tags/1.1.4/zopeapp.cfg: (404, 'Not Found')
Running /zfs/kveroneau/Sample/bin/buildout...
/bin/sh: 1: /zfs/kveroneau/Sample/bin/buildout: not found
```

After some research, I figured out that since zopetoolkit has moved over to GitHub, the URL redirects are now invalid and result in a 404.  The URL to fetch zopeapp.cfg should be:  https://raw.githubusercontent.com/zopefoundation/zopetoolkit/1.1.4/zopeapp.cfg

I find it really weird that this has also gone unnoticed in several years, really shows how many new projects are attempting to use Grok, or people try the tutorial and immediately fail and move on to the next web framework.

### PR DESCRIPTION
**Pull request by prsephton**
svn.zope.org disappeared, which broke our beloved grokproject.

This proposed change creates a 'versions' subfolder in the base directory which contains a fascimile of the original svn release directory.
```
 > ls versions/*
versions/current

versions/1.5.5:
versions.cfg

versions/eggs:
> cat versions/current 
1.5.5
```

The template is then changed to:
`GROK_BASE_URL = 'https://raw.githubusercontent.com/prsephton/'
GROK_RELEASE_URL = GROK_BASE_URL + 'grokproject/master/versions/'
`
Once merged, the GROK_BASE_URL can be changed to:
    https://raw.githubusercontent.com/zopefoundation/

For obvious quite a few tests broke due to this patch, since they expect to see "grok.zope.org" in the text.

Fixes #2

### CODE REVIEW

**Comment by prsephton** [comment]
All tests currently succeed on a freshly checked out copy of grokproject.  I have no clue what travis's issue is.

**Comment by icemac** [comment]
Thank you for working on this. Do you have a clue why TravisCI still fails?

**Comment by prsephton** [comment]
It has something to do with the version of setuptools used by the
buildout.  If you *sudo apt-get remove python-setuptools* and run
/grokproject/bin/buildout then the tests succeed.  If you *sudo apt-get
install python-setuptools* and rerun bin/buildout, it fails.  There is some
clash with a system module installed by setuptools which makes it fail.

Using virtualenv does not fix the problem either.

Regardless of TravisCI, running bin/grokproject ~/myprj installs a
perfectly working grok environment, together with working tests.

It drove me bats, so I stopped working on it.

**Comment by icemac** [comment]
@prsephton Other ZF projects install `setuptools` and `zc.buildout` via `pip` instead of using `bootstrap.py`, see [Zope](https://github.com/zopefoundation/Zope/blob/master/.travis.yml#L9-L13) as an example.

It is important that the versions installed via `pip` match the ones required in the `versions` section of `buildout.cfg` resp. the file it extends.

**Comment by prsephton** [comment]
Oh.  wow!  thanks.

I will go check.  However, current status is that the
grokproject/bin/buildout works perfectly for me (Python 2.7.9) regardless
of setuptools install status.  Adding a include-site-packages = false to
buildout.cfg seems to have fixed something.

Unfortunately, Travis still barfs, but with a new error this time.
[image: Inline image 1]

... which confuses me greatly, since I do not get this error from a clean
install on my dev machine.  There's clearly some version spec missing from
[versions], probably for zope-testrunner.  I will be looking into it just
as soon as I get a round toit.

Thanks for your help!

**Comment by prsephton** [comment]
Getting there.  No more version conflicts.  We are now failing on TravisCI because:

a)  pkg_resources/__init__.py gives a warning that makes expected text match fail for the test, and
b) https://raw.githubusercontent.com/zopefoundation/grokproject/master/versions/1.5.5/versions.cfg does not yet exist.  Obviously, since the PR contains that folder structure update to the github site.

It's possible that the __init__.py warning is stderr rather than stdout in which case it might not be a culprit.
_____________________________________________________________
 /home/travis/virtualenv/python2.7.13/lib/python2.7/site-packages/pkg_resources/__init__.py:184: RuntimeWarning: You have iterated over the result of pkg_resources.parse_version. This is a legacy behavior which is inconsistent with the new version class introduced in setuptools 8.0. In most cases, conversion to a tuple is unnecessary. For comparison of versions, sort the Version instances directly. If you have another use case requiring the tuple, please file a bug with the setuptools project describing that need.
...
    Error: Error downloading extends for URL https://raw.githubusercontent.com/zopefoundation/grokproject/master/versions/1.5.5/versions.cfg: (404, 'Not Found')

**Comment by sallner** [comment]
@prsephton I guess the clean solution would be to depend on the file in your PR-branch and then make a second PR that changes the file to the one on the master branch. It seems kind of double work, but helps to test before the merge.

**Comment by prsephton** [comment]
@sallner Actually, I was rather surprised that the download referenced _https://raw.githubusercontent.com/zopefoundation_ rather than _https://raw.githubusercontent.com/prsephton_, as the URL is hard coded in templates.py.  I assumed that github/TravisCI was doing something clever and replacing the site path.  I certainly am not purposely referencing _github/zopefoundation/grokproject_ in the code.

@icemac Using _pip install buildout_ and then running _buildout bootstrap/install_ causes problems with grokproject:  
grokproject (a) is a project which creates a project (b).   The tests are first run for grokproject (a) itself, and then for the temporary project (b) it creates.  

TravisCI runs inside a virtualenv, so the version of buildout installed by pip becomes available to both grokproject (a) as well as the created project (b).  This causes a library conflict when tests are run for (b).

If I set grokproject to use buildout 1.5.2, I get 
```
        raise VersionConflict(dist,req) # XXX put more info here
    pkg_resources.VersionConflict: (zc.buildout 1.5.2 (<virtual env path>/lib/python2.7/site-packages), Requirement.parse('zc.buildout==1.7.1'))
```

If I install version buildout 1.7.1 for grokproject (the version used by the installed project)
```
    Error: There is a version conflict.
    We already have: zc.buildout 1.7.1
```

Normally buildout is not installed into the system path, as is the case for _pip install buildout_, so in this case using _pip install buildout_ is not a true reflection of an actual runtime environment.

**Comment by icemac** [comment]
@prsephton Wow, you got it working, congratulations! 🥇 

@tseaver What is the current status of signing contributor agreements? http://foundation.zope.org still seems to be down. Or do you think we can merge this PR without requiring a contributor agreement?

**Comment by clozinski** [comment]
Contributor Agreements

I was recently at a talk by a lawyer on contributor agreements with regards to investors. 

While we developers may not care much about contributor agreements, investors care hugely.  During the due diligence process they check everything, and if the contributor agreements are not in place, that could literally affect the entire investment. 

I know that Zope corporation used to have a person responsible for managing this.  That is expensive. So what do we do?

I propose that we have a   GitHub repository for Zope Contributor Agreements. 

In the US, I believe that scans of signatures are legally valid.  Particularly when supported by the historic record.  But I am not a lawyer. 

Of course the law depends on the country.

**Comment by jcbrand** [comment]
On the topic of contributor agreements, I've been thoroughly impressed with [CLA Assistant](https://cla-assistant.io/) as used by the the XMPP Standards Foundation (https://github.com/xsf) on Github.

You sign and verify the agreement by proving ownership of your Github account (via OAuth).

It took me less than a minute to sign the contributor agreement. Would be great if Zope could also make it this easy.

**Comment by prsephton** [comment]
From my understanding I only need to be a registered committer to be the
person _merging_ the PR into the main repo?  Workflow is a bit different
from when the source was all in an SVN repo after all.

Personally, I don't care about legalities and only care about "Making Zope
Great Again"!!!

:-)

**Comment by clozinski** [comment]
>  Personally, I don't care about legalities and only care about "Making Zope
> Great Again"!!!

Well you are doing a good job of that.  What do you need help with?  

One comment.  The Zope brand name is totally botched, even more with Zope 4.  A good idea is to focus on Grok. And his logo is cuter.   In any case the grok permissions on views model is better than the more complex Zope permissions on objects model.

**Comment by prsephton** [comment]
I think we need to consider carefully and critically why you say the Zope
brand is botched.  I can think of a few reasons:

   - Zope Foundation no longer exists, along with it's clout regarding
   quality control and vision.
   - Python devs (some if not most) have no clue about Zope, and have heard
   negative things, and spread them.
   - No champions left (ok theres a few of us).  Maintainers don't seem to
   care about projects they used to maintain **
   - Lack of direction, communication and buy-in from devs (eg. Zope-4;
   what is it really?  Does anyone other than its devs care?)
   - Python 3 compatibility issues make the ZTK and Grok toolkit look
   unstable and outdated ***
   - Very low level of support.  ++
   - After learning Django, which other devs claim is easy and the best
   there is, the prospect of learning Zope, which everyone claims is hard and
   clunky is unattractive to most devs.
   - ZCML for configuration is not everyone's biscuit.
   - Too many breakages due to inattention and careless attitude
   substantiate claims of clunkiness.  $$
   - Other brands (nodejs/angularjs/etc) drown out the Zope brand.
   - Some really time consuming and confusing bugs make the dev environment
   too complicated. Especially for newcomers. ^^

** For example Grok's Martijn Faassen moved on a few years ago. Although he
still supports the principles in Grok, he replaced the ZCA with his own
registry when he built Morepath, and dropped traversal in favour of a
routing system- although his routes lead to objects and not views.
*** For example, megrok.rdb, an amazing wrapper over the equally amazing
zc.sqlalchemy does not even have a github entry.
++ Grok web site was incredibly slow for ages, and zopefoundation site is
still down.  It takes forever to see a response for some issues.
$$ For example, my recent issue raised with Grok not running due to
deprecation of zope.app.dublincore.permission in favor of
zope.dublincore.permission.  The deprecation happened in a dependent
module, and tests in that module ran just fine, but who knows what else
that change broke?  Since no one is actively developing Grok, the problem
went unnoticed.
^^ For example, buildout 2.x messes up some things really badly, so people
try to make do with virtualenv AND buildout together, leading to stupid
errors like egg version clashes between eggs installed in env/lib via pip
and those installed with buildout.  The current Grok TravisCI script, as
well as zopetoolkit is an example of this ridiculous approach. Either use
buildout or virtualenv; if buildout cannot provide the kind of dev/install
environment you need, drop it.

When I develop using the Grok toolkit, I know I am developing using the
Zope toolkit as well.  I have critisised the Zope brand myself, but it's no
worse than, for example, Pyramid.  At least you can search google for Zope
and get relevant stuff, while "Grok" often comes up with unrelated
articles, and Pyramid tells me about Egypt.

I agree that the logo is cuter.  Faassen still uses it as his twitter pic.
I probably would not be using Zope/ZTK at all were it not for Grok.

I am a HUGE fan of the original Zope/Grok devs.  Jan-Wybrand Kolman is
still around, and did some nice work on Python3 port, but people (even
devs) have a life too.  If Faassen wants to spend his time growing daisies
and vegetables, kudos to him!  He's retained his sanity AND enjoying life.
Nevertheless, we need some sort of direction, control, support and above
all, champions!  If we still had a zopefoundation, we might have been able
to retain some of these heavyweights in advisory roles.

Regards,

PS: on a side note, one of those things buildout-2 does worse than
buildout-1 is interpreter script creation.  Buildout-2 creates an unusable
zc.buildout.egg:scripts interpreter, and by unusable I mean you cannot use
it to configure pycharm/eclipse projects.  I have a solution for that (with
buiildout-2) if you are looking for one.  Buildout-1 creates usable
interpreters out of the box with z3c.buildout.scripts recipe but that no
longer works in buildout-2.  One of those instances of progress on closer
investigation being regression.

**Comment by icemac** [comment]
@prsephton wrote:
> Zope Foundation no longer exists, along with it's clout regarding quality control and vision.

It is the Zope Corporation which no longer exists. Zope Foundation is still existing.

**Comment by prsephton** [comment]
Oh, right, thanks! my bad!

**Comment by prsephton** [comment]
What is stopping this merge from happening?

**Comment by clozinski** [comment]
I propose that the powers that be give Paul Sephton permission to accept pull requests on GrokProject, and any other grok repositories which he wants to work on.

**Comment by prsephton** [comment]
I suspect the problem is that the powers that be are off planet at the
moment :-)

**Comment by jcbrand** [comment]
@janwijbrand, @janjaapdriessen or @thefunny42 could any of you perhaps give @prsephton permission to accept pull requests on "grokproject", or at least just merge this?

**Comment by clozinski** [comment]
That would be lovely if someone promoted Paul.  He deserves it. 

Then I could announce a new version of GrokProject at my ZODB Ecosystem talk at PyCON UK later this week.

**Comment by icemac** [comment]
@prsephton @clozinski This PR seems like a substantial change in this repository. The default policy is that the one who wrote the code merges the PR. 

The way to get the right to merge PRs and to write to the zopefoundation repositories is to sign a contributor agreement: https://docs.zope.org/developer/becoming-a-committer.html Contributors get write access to all zopefoundation repositories including the grok ones. (The link is live again, I double checked it.)

**Comment by janwijbrand** [comment]
I agree on giving @prsephton write access. If this means signing the agreement, then let's try and help him in getting this signed. The grok project developers were indeed somewhat "off the planet". It is cool to see there is still progress. Thank you for that.

**Comment by janwijbrand** [comment]
Of course, if you want us too, we could merge the PR quickly (and perhaps even release?)

**Comment by prsephton** [comment]
Thanks, guys, the paperwork is submitted.

@janwijbrand You are welcome to merge if you like, please just check the version tags.  Otherwise I can merge once the contributor process has completed (no idea how long that takes).

**Comment by prsephton** [comment]
Oh, btw I didn't mean to close this :-)

File: tests.txt [hunk 1]
```diff
@@ -11,14 +11,26 @@ Eggs are placed in our freshly created eggs directory:
...
     >>> # point in the future.
     >>> executable = [os.path.join(current_dir, 'bin', 'grokproject')]
+
+When we have merged the PR, this will need to change
+
+    >>> base = open(os.path.join(current_dir, '.git', 'config')).read()
+    >>> base = base[base.find('url = ')+6:base.find('.git')]
+    >>> base = base.replace('github.com', 'raw.githubusercontent.com')
+
+We temporarily hard code 'prsephton/grokproject' since the version info does not
```
Comment (janwijbrand) [inline tests.txt]: Do you mean you'll remove the hardcoded paths after this PR is merged? And will that removal then be in its own PR?

**Comment by prsephton** [comment]
Oh, and by the way I'm trying to figure out why Travis could not download PasteScript==2.0.2.

File: tests.txt (hunk 1 above)
Comment (prsephton) [reply tests.txt]: Yes.  The first PR will be using the 'github/prsephton/grokproject' repo for a 'versions' directory.  Once merged we should have a 'versions' dir in zopefoundation/grokproject, and the second PR will fix the code to use that. 

The rain dance is just to ensure that things are never in a broken state and that the code can always reach a working 'versions' directory.

**Comment by prsephton** [comment]
Looks like that was a glitch in travis.  Build is successful in my clone of the repo.  Closing and re-opening should trigger another travis build which will hopefully succeed this time.

**Comment by prsephton** [comment]
Re-open for travis trigger.

File: buildout.cfg [hunk 2]
```diff
@@ -1,6 +1,13 @@
 [buildout]
 newest = false
+include-site-packages = false
 develop = .
+#extensions = buildout.dumppickedversions
+#dump-picked-versions-file = versions.cfg
+#overwrite-picked-versions-file = false
```
Comment (tseaver) [inline buildout.cfg]: Please remove the commented-out lines.

File: grokproject/templates.py [hunk 3]
```diff
@@ -1,7 +1,16 @@
...
+try:
+    import urllib2
+    import urlparse
+    from urlparse import urljoin
+    from urllib2 import HTTPError, urlopen
+except ImportError:
+    import urllib as urllib2
+    from urllib.parse import urlparse, urljoin
+    from urllib.error import HTTPError
+    from urllib.request import urlopen
```
Comment (tseaver) [inline grokproject/templates.py]: Might be better handled through `six.moves`?

Comment (prsephton) [reply grokproject/templates.py]: Indeed.  Thanks for the pointers.  Done.

**Comment by prsephton** [comment]
Merged changes.  Nice to have a working grokproject again.

**Comment by tseaver** [comment]
Welcome aboard, and thanks for your efforts!

**Comment by prsephton** [comment]
Thanks, Tres :-)
I hope I can be of some use.  I will probably need some hand-holding/clouting initially until I find the ropes.
